src/BatteryStateOfChargeService.cpp

```diff
@@ -43,21 +48,55 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    if (totalAH_>BATTERY_AMP_HOUR_CAPACITY)
+        return 0;
+    else if (totalAH_<0)
+        return BATTERY_AMP_HOUR_CAPACITY;
+    else
+        return BATTERY_AMP_HOUR_CAPACITY - totalAH_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return isCharging_;
 }
 
+
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime null_time(0,0,0,0);
+    if (return_time.isNull())
+        return null_time;
+    else
+        return return_time;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+    if (previousTime_.isNull())
```

Comment:
Really good! Out of all the times I've seen how people checked for the first time this function has been ran, this is probably the best solution (better than my solution! I didn't do this).